src/realm/table.cpp

```diff
@@ -1692,14 +1722,78 @@ void Table::upgrade_olddatetime()
 }
 
 
-void Table::add_search_index(size_t col_ndx)
+void Table::add_search_index(size_t col_ndx, DescriptorRef* subdesc)
 {
     if (REALM_UNLIKELY(!is_attached()))
         throw LogicError(LogicError::detached_accessor);
 
-    if (REALM_UNLIKELY(has_shared_type()))
-        throw LogicError(LogicError::wrong_kind_of_table);
+    if (subdesc) {
+        // Add index to column of subtable given in subdesc. `this` table must be its root table
+        if (!get_descriptor()->is_root())
+            throw(LogicError::wrong_kind_of_table);
+
+        typedef _impl::DescriptorFriend df;
+        size_t tmp;
+
+        // Find the root table column index that contains the search index
+        size_t* parent_col = df::record_subdesc_path(*subdesc->get(), &tmp, &tmp + 1);
+        if (!parent_col) {
+            throw LogicError(LogicError::subtable_of_subtable_index);
+        }
+        
+        if (REALM_UNLIKELY(col_ndx >= subdesc->get()->get_spec()->get_column_count()))
+            throw LogicError(LogicError::column_index_out_of_range);
+
+        // Early-out of already indexed
+        int attr = subdesc->get()->get_spec()->get_column_attr(col_ndx);
+        if (attr & col_attr_Indexed)
+            return;
+
+        // Iterate through all rows of the root table and create an instance of each subtable. We have a special
+        // problem though: All subtables share the same common instance of attributes, however while we successively
+        // create indexes, we have some subtables that will match the index flag and some that will not, which
+        // is an inchoherent state of the database. We rely on the fact that all method calls and operations in the 
+        // for-loop are safe to call despite of this.
```

Comment:
Nice comment!

Would it be correct to say that this inconsistent state is fine temporarily, because if it should happen that anything throws in the body of the loop, the current write transaction will be rolled back, so the inconsistent state will never "leak" beyond `commit()` anyway?